src/qengine/state/opencl.cpp

```diff
@@ -106,25 +114,84 @@ void QEngineOCL::Apply2x2(bitCapInt offset1, bitCapInt offset2, const Complex16*
     queue.enqueueUnmapMemObject(stateBuffer, stateVec);
     queue.enqueueWriteBuffer(cmplxBuffer, CL_FALSE, 0, sizeof(Complex16) * CMPLX_NORM_LEN, cmplx);
     queue.enqueueWriteBuffer(ulongBuffer, CL_FALSE, 0, sizeof(bitCapInt) * BCI_ARG_LEN, bciArgs);
+    if (doCalcNorm) {
+        nrmParts = new double[CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE]();
+        queue.enqueueWriteBuffer(nrmBuffer, CL_FALSE, 0, sizeof(double) * CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE, nrmParts);
+    }
     queue.finish();
 
-    cl::Kernel apply2x2 = *(clObj->GetApply2x2Ptr());
+    cl::Kernel apply2x2;
+    if (doCalcNorm) {
+         apply2x2 = *(clObj->GetApply2x2NormPtr());
+    } else {
+         apply2x2 = *(clObj->GetApply2x2Ptr());
+    }
     apply2x2.setArg(0, stateBuffer);
     apply2x2.setArg(1, cmplxBuffer);
     apply2x2.setArg(2, ulongBuffer);
+    if (doCalcNorm) {
+        apply2x2.setArg(3, nrmBuffer);
+    }
     queue.enqueueNDRangeKernel(apply2x2, cl::NullRange, // kernel, offset
-        cl::NDRange(CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE), // global number of work items
-        cl::NDRange(1)); // local number (per group)
+            cl::NDRange(CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE), // global number of work items
+            cl::NDRange(1)); // local number (per group)
 
-    queue.enqueueMapBuffer(stateBuffer, CL_TRUE, CL_MAP_READ | CL_MAP_WRITE, 0, sizeof(Complex16) * maxQPower);
+    queue.enqueueMapBuffer(stateBuffer, CL_TRUE, CL_MAP_WRITE, 0, sizeof(Complex16) * maxQPower);
     if (doCalcNorm) {
-        UpdateRunningNorm();
+        queue.enqueueReadBuffer(nrmBuffer, CL_TRUE, 0, sizeof(double) * CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE, nrmParts);
+        runningNorm = 0.0;
+        for (unsigned long int i = 0; i < CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE; i++) {
+            runningNorm += nrmParts[i];
+        }
+        delete[] nrmParts;
+        runningNorm = sqrt(runningNorm);
     } else {
         runningNorm = 1.0;
     }
 
 }
 
+void QEngineOCL::X(bitLenInt qubit)
```

Comment:
So... remove?

Reply:
This tripped me up. It's actually required based on the rules of inheritance, because of how we're overloading X(), as far as I can tell. I only added this in the first place because of an error without it. The other option might be a default length parameter of 1, but that mucks up the interface, which doesn't know about these length=1 cases.